src/cell.rs

```diff
@@ -90,81 +90,261 @@ pub fn range_manhattan4<T: Float>(p1: math::Point4<T>, p2: math::Point4<T>) -> T
+    let r0101 = range_func(*point, p0101);
+    let r1101 = range_func(*point, p1101);
+    let r0011 = range_func(*point, p0011);
+    let r1011 = range_func(*point, p1011);
+    let r0111 = range_func(*point, p0111);
+    let r1111 = range_func(*point, p1111);
+
+    if r0000 < r1000 && r0000 < r0100 && r0000 < r1100 && r0000 < r0010 && r0000 < r1010 && r0000 < r0110 && r0000 < r1110 &&
+        r0000 < r0001 && r0000 < r1001 && r0000 < r0101 && r0000 < r1101 && r0000 < r0011 && r0000 < r1011 && r0000 < r0111 && r0000 < r1111 {
+        (p0000, r0000)
+    } else if r1000 < r0100 && r1000 < r1100 && r1000 < r0010 && r1000 < r1010 && r1000 < r0110 && r1000 < r1110 && r1000 < r0001 &&
+        r1000 < r1001 && r1000 < r0101 && r1000 < r1101 && r1000 < r0011 && r1000 < r1011 && r1000 < r0111 && r1000 < r1111 {
+        (p1000, r1000)
+    } else if r0100 < r1100 && r0100 < r0010 && r0100 < r1010 && r0100 < r0110 && r0100 < r1110 && r0100 < r0001 && r0100 < r1001 &&
+        r0100 < r0101 && r0100 < r1101 && r0100 < r0011 && r0100 < r1011 && r0100 < r0111 && r0100 < r1111 {
+        (p0100, r0100)
+    } else if r1100 < r0010 && r1100 < r1010 && r1100 < r0110 && r1100 < r1110 && r1100 < r0001 && r1100 < r1001 && r1100 < r0101 &&
+        r1100 < r1101 && r1100 < r0011 && r1100 < r1011 && r1100 < r0111 && r1100 < r1111 {
+        (p1100, r1100)
+    } else if r0010 < r1010 && r0010 < r0110 && r0010 < r1110 && r0010 < r0001 && r0010 < r1001 && r0010 < r0101 && r0010 < r1101 &&
+        r0010 < r0011 && r0010 < r1011 && r0010 < r0111 && r0010 < r1111 {
+        (p0010, r0010)
+    } else if r1010 < r0110 && r1010 < r1110 && r1010 < r0001 && r1010 < r1001 && r1010 < r0101 && r1010 < r1101 && r1010 < r0011 &&
+        r1010 < r1011 && r1010 < r0111 && r1010 < r1111 {
+        (p1010, r1010)
+    } else if r0110 < r1110 && r0110 < r0001 && r0110 < r1001 && r0110 < r0101 && r0110 < r1101 && r0110 < r0011 && r0110 < r1011 &&
+        r0110 < r0111 && r0110 < r1111 {
+        (p0110, r0110)
+    } else if r1110 < r0001 && r1110 < r1001 && r1110 < r0101 && r1110 < r1101 && r1110 < r0011 && r1110 < r1011 && r1110 < r0111 &&
+        r1110 < r1111 {
+        (p1110, r1110)
+    } else if r0001 < r1001 && r0001 < r0101 && r0001 < r1101 && r0001 < r0011 && r0001 < r1011 && r0001 < r0111 && r0001 < r1111 {
+        (p0001, r0001)
+    } else if r1001 < r0101 && r1001 < r1101 && r1001 < r0011 && r1001 < r1011 && r1001 < r0111 && r1001 < r1111 {
+        (p1001, r1001)
+    } else if r0101 < r1101 && r0101 < r0011 && r0101 < r1011 && r0101 < r0111 && r0101 < r1111 {
+        (p0101, r0101)
+    } else if r1101 < r0011 && r1101 < r1011 && r1101 < r0111 && r1101 < r1111 {
+        (p1101, r1101)
+    } else if r0011 < r1011 && r0011 < r0111 && r0011 < r1111 {
+        (p0011, r0011)
+    } else if r1011 < r0111 && r1011 < r1111 {
+        (p1011, r1011)
+    } else if r0111 < r1111 {
+        (p0111, r0111)
+    } else {
+        (p1111, r1111)
+    }
```

Comment:
I am unrolling the loop... But the important part is that I'm checking a lot fewer points.

The classic Worley approach to cell noise is to place a seed point inside each cell, and then check the seed points in the current cell, and all neighboring cells. This produces a progression of 3^n, where n is the number of dimensions. That is, 9-27-81 for 2D, 3D and 4D.

I changed it to work more like Perlin noise. It places a seed point in the vicinity of each cell corner, and checks against all corners of the current cell. This produces a progression of n^2, or 4-8-16 for 2D, 3D and 4D.

I think the only reason this didn't produce a much better speed-up is the fact that it now needs a gradient table and a bit of extra math, because in order to avoid discontinuities it needs to place the points in a hypersphere instead of a hypercube.